include/spi.h

```diff
@@ -27,173 +37,182 @@
- *    transfer_mode  [ 3 ]       - LSB or MSB first transfer mode.
- *    word_size      [ 4 : 11 ]  - Size of a data frame in bits.
- *    RESERVED       [ 12 : 31 ] - Undefined or device-specific usage.
- *
- * max_sys_freq is the clock divider supported by the the host
- * spi controller.
+ * @brief SPI MISO lines
  */
-struct spi_config {
-	u32_t	config;
-	u32_t	max_sys_freq;
-};
+#define SPI_LINES_SINGLE	(0)
+#define SPI_LINES_DUAL		BIT(11)
+#define SPI_LINES_QUAD		BIT(12)
 
-struct spi_buf {
+#define SPI_LINES_MASK		(0x3 << 11)
 
+/**
+ * @brief SPI Chip Select control structure
+ *
+ * gpio_dev is a valid pointer to an actual GPIO device
+ * gpio_pin is a number representing the gpio PIN that will be used
+ *    to act as a CS line
+ * delay is a delay in microseconds to wait before starting the
+ *    transmission and before releasing the CS line
+ */
+struct spi_cs_control {
+	struct device	*gpio_dev;
+	u32_t		gpio_pin;
+	u32_t		delay;
 };
 
 /**
- * @typedef spi_api_configure
- * @brief Callback API upon configuring the const controller
- * See spi_configure() for argument description
+ * @brief SPI controller configuration structure
+ *
+ * frequency is the bus frequency in Hertz
+ * operation is a bit field with the following parts:
+ *    operational mode    [ 0 ]       - master or slave.
+ *    mode                [ 1 : 3 ]   - Polarity, phase and loop mode.
+ *    transfer            [ 4 ]       - LSB or MSB first.
+ *    word_size           [ 5 : 10 ]  - Size of a data frame in bits.
+ *    lines               [ 11 : 12 ] - MISO lines: Single/Dual/Quad.
+ *    RESERVED            [ 13 : 15 ] - Unused yet
+ *
+ * slave is the slave number from 0 to host constoller slave limit.
+ *
+ * cs is a valid pointer on a struct spi_cs_control is CS line is
+ *    emulated through a gpio line, or NULL otherwise.
  */
-typedef int (*spi_api_configure)(struct device *dev,
-				 struct spi_config *config);
+struct spi_config {
+	u32_t		frequency;
+	u16_t		operation;
+	u16_t		slave;
+	struct spi_cs_control *cs;
+};
+
 /**
- * @typedef spi_api_slave_select
- * @brief Callback API upon selecting a slave
- * See spi_slave_select() for argument description
+ * @brief SPI buffer structure
+ *
+ * buf is a valid pointer on a data buffer, or NULL otherwise.
+ * len is the length of the buffer or, if buf is NULL, will be the
+ *     length which as to be sent as dummy bytes (as TX buffer) or
+ *     the length of bytes that should be skipped (as RX buffer).
  */
-typedef int (*spi_api_slave_select)(struct device *dev, u32_t slave);
+struct spi_buf {
+	void *buf;
+	u32_t len;
+};
+
 /**
  * @typedef spi_api_io
  * @brief Callback API for I/O
- * See spi_read() and spi_write() for argument descriptions
+ * See spi_transceive() for argument descriptions
  */
-typedef int (*spi_api_io)(struct device *dev,
-			  const void *tx_buf, u32_t tx_buf_len,
-			  void *rx_buf, u32_t rx_buf_len);
+typedef int (*spi_api_io)(struct device	*dev,
```

Comment:
It seems a bit harder to tease apart the errors due to misconfiguration from I/O errors here than in the old API. Will you specify which errno values correspond to which?

In general, what is the rationale for removing the separate configuration routine?

This seems to add some overhead, as the SPI controller must re-check the configuration each time its spi_api_io is called, rather than being able to kick off new I/O with an already established configuration. This seems like it could decrease throughput in back-to-back transactions.